In [50]:
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.tree import export_graphviz
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix


## Exercises

### 1. Create a new file named model_evaluation.py or model_evaluation.ipynb for these exercises.
### 2. Given the following confusion matrix, evaluate (by hand) the model's performance.

|               | pred dog   | pred cat   |
|:------------  |-----------:|-----------:|
| actual dog    |         46 |         7  |
| actual cat    |         13 |         34 |




### a. In the context of this problem, what is a false positive?

Let's say cat is positive class and dog is negative class.

False Posistive = 7 (cat is predicted but it actually is dog)

### b. In the context of this problem, what is a false negative?

False Negative = 13 (dog is predicted but it actually is cat)

### c. How would you describe this model?


Accuracy: (TP + TN) / (TP + TN + FP + FN)

In [2]:
(34 + 46) / (34 + 46 + 7 + 13)

0.8

Precision = (TP) / (TP + FP)

In [3]:
34 / (34 + 7)

0.8292682926829268

It's Cat-classifier/model with 80% accuracy and 83% precision since cat is the positive prediction.

### 3. You are working as a datascientist working for Codeup Cody Creator (C3 for short), a rubber-duck manufacturing plant.

### Unfortunately, some of the rubber ducks that are produced will have defects. Your team has built several models that try to predict those defects, and the data from their predictions can be found here.

### Use the predictions dataset and pandas to help answer the following questions:


In [4]:
df = pd.read_csv('c3.csv')
df.head()

actual     model1  model2     model3
0  No Defect  No Defect  Defect  No Defect
1  No Defect  No Defect  Defect     Defect
2  No Defect  No Defect  Defect  No Defect
3  No Defect     Defect  Defect     Defect
4  No Defect  No Defect  Defect  No Defect

### An internal team wants to investigate the cause of the manufacturing defects. They tell you that they want to identify as many of the ducks that have a defect as possible. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?



Let's say, Positive class : ducks have a defect 

Negative class: ducks have NO defect

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   actual  200 non-null    object
 1   model1  200 non-null    object
 2   model2  200 non-null    object
 3   model3  200 non-null    object
dtypes: object(4)
memory usage: 6.4+ KB


In [6]:
# Finding class
df.actual.value_counts()

No Defect    184
Defect        16
Name: actual, dtype: int64

In [7]:
# Model 1

subset = df[df.actual == 'Defect']
subset

actual     model1     model2     model3
13   Defect  No Defect     Defect     Defect
30   Defect     Defect  No Defect     Defect
65   Defect     Defect     Defect     Defect
70   Defect     Defect     Defect     Defect
74   Defect  No Defect  No Defect     Defect
87   Defect  No Defect     Defect     Defect
118  Defect  No Defect     Defect  No Defect
135  Defect     Defect  No Defect     Defect
140  Defect  No Defect     Defect     Defect
147  Defect     Defect  No Defect     Defect
163  Defect     Defect     Defect     Defect
171  Defect  No Defect     Defect     Defect
176  Defect  No Defect     Defect     Defect
186  Defect  No Defect  No Defect  No Defect
194  Defect     Defect  No Defect     Defect
196  Defect     Defect  No Defect  No Defect

In [8]:
# Finding Model 1 recall

model_recall = (subset.actual == subset.model1).mean()
print("Model 1")
print(f"Model recall: {model_recall:.2%}")

Model 1
Model recall: 50.00%


In [9]:
# Finding Model 2 recall

model_recall = (subset.actual == subset.model2).mean()
print("Model 2")
print(f"Model recall: {model_recall:.2%}")


Model 2
Model recall: 56.25%


In [10]:
# Finding Model 3 recall

model_recall = (subset.actual == subset.model3).mean()
print("Model 3")
print(f"Model recall: {model_recall:.2%}")

Model 3
Model recall: 81.25%


Recall evaluation metric would be appropriate here.

Model 3 would be the best fit as it has the highest recall of 81.25%


### Recently several stories in the local news have come out highlighting customers who received a rubber duck with a defect, and portraying C3 in a bad light. The PR team has decided to launch a program that gives customers with a defective duck a vacation to Hawaii. They need you to predict which ducks will have defects, but tell you the really don't want to accidentally give out a vacation package when the duck really doesn't have a defect. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?



In [11]:
# for model 1 
subset = df[df.model1 == 'Defect']
subset

actual  model1     model2     model3
3    No Defect  Defect     Defect     Defect
30      Defect  Defect  No Defect     Defect
62   No Defect  Defect  No Defect  No Defect
65      Defect  Defect     Defect     Defect
70      Defect  Defect     Defect     Defect
135     Defect  Defect  No Defect     Defect
147     Defect  Defect  No Defect     Defect
163     Defect  Defect     Defect     Defect
194     Defect  Defect  No Defect     Defect
196     Defect  Defect  No Defect  No Defect

In [12]:
model1_precision = (subset.actual == subset.model1).mean()
print(f'model1 precision: {model1_precision:.2%}')

model1 precision: 80.00%


In [13]:
# for model 2
subset = df[df.model2 == 'Defect']
subset



actual     model1  model2     model3
0    No Defect  No Defect  Defect  No Defect
1    No Defect  No Defect  Defect     Defect
2    No Defect  No Defect  Defect  No Defect
3    No Defect     Defect  Defect     Defect
4    No Defect  No Defect  Defect  No Defect
..         ...        ...     ...        ...
183  No Defect  No Defect  Defect  No Defect
185  No Defect  No Defect  Defect  No Defect
193  No Defect  No Defect  Defect     Defect
195  No Defect  No Defect  Defect     Defect
198  No Defect  No Defect  Defect     Defect

[90 rows x 4 columns]

In [14]:
model2_precision = (subset.actual == subset.model2).mean()
print(f'model2 precision: {model2_precision:.2%}')

model2 precision: 10.00%


In [15]:
# for model 3
subset = df[df.model3 == 'Defect']
subset

actual     model1     model2  model3
1    No Defect  No Defect     Defect  Defect
3    No Defect     Defect     Defect  Defect
5    No Defect  No Defect  No Defect  Defect
9    No Defect  No Defect  No Defect  Defect
13      Defect  No Defect     Defect  Defect
..         ...        ...        ...     ...
193  No Defect  No Defect     Defect  Defect
194     Defect     Defect  No Defect  Defect
195  No Defect  No Defect     Defect  Defect
198  No Defect  No Defect     Defect  Defect
199  No Defect  No Defect  No Defect  Defect

[99 rows x 4 columns]

In [16]:
model3_precision = (subset.actual == subset.model3).mean()
print(f'model3 precision: {model3_precision:.2%}')

model3 precision: 13.13%


Model 1 need to be used because it helps to lower the false predictions of defects.

### 4. You are working as a data scientist for Gives You Paws ™, a subscription based service that shows you cute pictures of dogs or cats (or both for an additional fee).

### At Gives You Paws, anyone can upload pictures of their cats or dogs. The photos are then put through a two step process. First an automated algorithm tags pictures as either a cat or a dog (Phase I). Next, the photos that have been initially identified are put through another round of review, possibly with some human oversight, before being presented to the users (Phase II).

### Several models have already been developed with the data, and you can find their results here.

### Given this dataset, use pandas to create a baseline model (i.e. a model that just predicts the most common class) and answer the following questions:




Creating a baseline model


In [17]:
paws = pd.read_csv('gives_you_paws.csv')
paws.head()

actual model1 model2 model3 model4
0    cat    cat    dog    cat    dog
1    dog    dog    cat    cat    dog
2    dog    cat    cat    cat    dog
3    dog    dog    dog    cat    dog
4    cat    cat    cat    dog    dog

In [18]:
paws.actual.value_counts()

dog    3254
cat    1746
Name: actual, dtype: int64

In [19]:
# The most frequent label is dog.
paws['baseline_prediction'] = 'dog'

In [20]:
paws.head()

actual model1 model2 model3 model4 baseline_prediction
0    cat    cat    dog    cat    dog                 dog
1    dog    dog    cat    cat    dog                 dog
2    dog    cat    cat    cat    dog                 dog
3    dog    dog    dog    cat    dog                 dog
4    cat    cat    cat    dog    dog                 dog

 ###  a. In terms of accuracy, how do the various models compare to the baseline model? Are any of the models better than the baseline?

In [21]:
# model 1
model1_accuracy = (paws.actual == paws.model1).mean()
baseline_accuracy = (paws.actual == paws.baseline_prediction).mean()

print(f'model1 accuracy: {model1_accuracy:.2%}')
print(f'baseline accuracy: {baseline_accuracy:.2%}')

model1 accuracy: 80.74%
baseline accuracy: 65.08%


In [22]:
# model 2 
model2_accuracy = (paws.actual == paws.model2).mean()
baseline_accuracy = (paws.actual == paws.baseline_prediction).mean()

print(f'model2 accuracy: {model2_accuracy:.2%}')
print(f'baseline accuracy: {baseline_accuracy:.2%}')

model2 accuracy: 63.04%
baseline accuracy: 65.08%


In [23]:
# model 3
model3_accuracy = (paws.actual == paws.model3).mean()
baseline_accuracy = (paws.actual == paws.baseline_prediction).mean()

print(f'model3 accuracy: {model3_accuracy:.2%}')
print(f'baseline accuracy: {baseline_accuracy:.2%}')

model3 accuracy: 50.96%
baseline accuracy: 65.08%


In [24]:
# model 4
model4_accuracy = (paws.actual == paws.model4).mean()
baseline_accuracy = (paws.actual == paws.baseline_prediction).mean()

print(f'model4 accuracy: {model4_accuracy:.2%}')
print(f'baseline accuracy: {baseline_accuracy:.2%}')

model4 accuracy: 74.26%
baseline accuracy: 65.08%


#### models 1 and 4 are better than the baseline.

###  b. Suppose you are working on a team that solely deals with dog pictures. Which of these models would you recomend for Phase I? For Phase II?

 phase 1

In [25]:
subset = paws[paws.actual == 'dog']
subset.head()

actual model1 model2 model3 model4 baseline_prediction
1    dog    dog    cat    cat    dog                 dog
2    dog    cat    cat    cat    dog                 dog
3    dog    dog    dog    cat    dog                 dog
5    dog    dog    dog    dog    dog                 dog
8    dog    dog    cat    dog    dog                 dog

In [26]:
# model 1 
(subset.actual == subset.model1).mean()

0.803318992009834

In [27]:
# model 2 
(subset.actual == subset.model2).mean()



0.49078057775046097

In [28]:
# model 3
(subset.actual == subset.model3).mean()

0.5086047940995697

In [29]:
# model 4 
(subset.actual == subset.model4).mean()

0.9557467732022127

In [30]:
# We know model 4 has the highest subset.
pd.crosstab(paws.model4, paws.actual)

actual   cat   dog
model4            
cat      603   144
dog     1143  3110

phase 2

In [31]:
# wer are creating subset for each model
subset1 = paws[paws.model1 == 'dog']
subset2 = paws[paws.model2 == 'dog']
subset3 = paws[paws.model3 == 'dog']
subset4 = paws[paws.model4 == 'dog']

In [32]:
# model 1 
(subset1.actual == subset1.model1).mean()

0.8900238338440586

In [33]:
# model 2
(subset2.actual == subset2.model2).mean()

0.8931767337807607

In [34]:
# model 3
(subset3.actual == subset3.model3).mean()

0.6598883572567783

In [35]:
# model 4 
(subset4.actual == subset4.model4).mean()

0.7312485304490948

In [36]:
# we know model 2 has the highest precision.
pd.crosstab(paws.model2, paws.actual)

actual   cat   dog
model2            
cat     1555  1657
dog      191  1597

Use model 4 for recall which is 96%.(from phase 1)

Use model 2 for precision which is 89%.(from phase 2)

### c. Suppose you are working on a team that solely deals with cat pictures. Which of these models would you recomend for Phase I? For Phase II?

we will do the same steps for cat this time as we did for the dog previously.

phase 1 

In [37]:
subset = paws[paws.actual == 'cat']
subset.head()

actual model1 model2 model3 model4 baseline_prediction
0     cat    cat    dog    cat    dog                 dog
4     cat    cat    cat    dog    dog                 dog
6     cat    cat    cat    cat    dog                 dog
7     cat    dog    cat    cat    dog                 dog
11    cat    cat    dog    cat    cat                 dog

In [38]:
# model 1 
(subset.actual == subset.model1).mean()

0.8150057273768614

In [39]:
# model 2 
(subset.actual == subset.model2).mean()

0.8906071019473081

In [40]:
# model 3
(subset.actual == subset.model3).mean()

0.5114547537227949

In [41]:
# model 4 
(subset.actual == subset.model4).mean()

0.34536082474226804

In [42]:
# we know that model 2 has the highest subset.
pd.crosstab(paws.model2, paws.actual)

actual   cat   dog
model2            
cat     1555  1657
dog      191  1597

phase 2

In [43]:
# creating subset for each model now.
subset1 = paws[paws.model1 == 'cat']
subset2 = paws[paws.model2 == 'cat']
subset3 = paws[paws.model3 == 'cat']
subset4 = paws[paws.model4 == 'cat']

In [44]:
# model 1 
(subset1.actual == subset1.model1).mean()

0.6897721764420747

In [45]:
# model 2
(subset2.actual == subset2.model2).mean()

0.4841220423412204

In [46]:
# model 3
(subset3.actual == subset3.model3).mean()

0.358346709470305

In [47]:
# model 4 
(subset4.actual == subset4.model4).mean()

0.8072289156626506

In [48]:
# we know that model 4 has the highest precision. 
pd.crosstab(paws.model4, paws.actual)

actual   cat   dog
model4            
cat      603   144
dog     1143  3110

Use model 2 for recall which is 89%. (phase 1)

Use model 4 for precision which is 81%. (phase 2)

### 5. Follow the links below to read the documentation about each function, then apply those functions to the data from the previous problem.

#### sklearn.metrics.accuracy_score
#### sklearn.metrics.precision_score
#### sklearn.metrics.recall_score
#### sklearn.metrics.classification_report

Let's say cat is positive class and dog is negative class.

In [51]:
x = classification_report(paws.actual, paws.model1, labels = ['cat', 'dog'], output_dict=True)



In [52]:
pd.DataFrame(x).T

precision    recall  f1-score    support
cat            0.689772  0.815006  0.747178  1746.0000
dog            0.890024  0.803319  0.844452  3254.0000
accuracy       0.807400  0.807400  0.807400     0.8074
macro avg      0.789898  0.809162  0.795815  5000.0000
weighted avg   0.820096  0.807400  0.810484  5000.0000

In [53]:
# model 1 
print("Model 1")
pd.DataFrame(classification_report(paws.actual, paws.model1, labels = ['cat', 'dog'], output_dict=True)).T

Model 1


precision    recall  f1-score    support
cat            0.689772  0.815006  0.747178  1746.0000
dog            0.890024  0.803319  0.844452  3254.0000
accuracy       0.807400  0.807400  0.807400     0.8074
macro avg      0.789898  0.809162  0.795815  5000.0000
weighted avg   0.820096  0.807400  0.810484  5000.0000

In [54]:
# model 2
print("Model 2")
pd.DataFrame(classification_report(paws.actual, paws.model2, labels = ['cat', 'dog'], output_dict=True)).T

Model 2


precision    recall  f1-score    support
cat            0.484122  0.890607  0.627269  1746.0000
dog            0.893177  0.490781  0.633479  3254.0000
accuracy       0.630400  0.630400  0.630400     0.6304
macro avg      0.688649  0.690694  0.630374  5000.0000
weighted avg   0.750335  0.630400  0.631310  5000.0000

In [55]:
# model 3
print("Model 3")
pd.DataFrame(classification_report(paws.actual, paws.model3, labels = ['cat', 'dog'], output_dict=True)).T


Model 3


precision    recall  f1-score    support
cat            0.358347  0.511455  0.421425  1746.0000
dog            0.659888  0.508605  0.574453  3254.0000
accuracy       0.509600  0.509600  0.509600     0.5096
macro avg      0.509118  0.510030  0.497939  5000.0000
weighted avg   0.554590  0.509600  0.521016  5000.0000

In [56]:
# model 4 
print("Model 4")
pd.DataFrame(classification_report(paws.actual, paws.model4, labels = ['cat', 'dog'], output_dict=True)).T

Model 4


precision    recall  f1-score    support
cat            0.807229  0.345361  0.483755  1746.0000
dog            0.731249  0.955747  0.828560  3254.0000
accuracy       0.742600  0.742600  0.742600     0.7426
macro avg      0.769239  0.650554  0.656157  5000.0000
weighted avg   0.757781  0.742600  0.708154  5000.0000